### ISSUE TITLE
Integer value out of range for enumeration type causes hard error rather than SFINAE

### ISSUE THREAD

**Issue by improbable-nickkrempel** [llvm/llvm-project#57176]
For example, compiled with clang trunk with `-std=c++20` ([godbolt](https://gcc.godbolt.org/z/PsYofabMb)):

```
#include <iostream>

enum Enum { A, B, C };

template <int V, Enum X = static_cast<Enum>(V)>
constexpr bool test(int) {
    return true;
}

template <int V>
constexpr bool test(...) {
    return false;
}

int main() {
    std::cout << test<4>(0) << '\n';
}
```

Results in an error:
```
<source>:7:27: error: integer value 4 is outside the valid range of values [0, 3] for this enumeration type [-Wenum-constexpr-conversion]
template <int V, Enum X = static_cast<Enum>(V)>
```

(Adding the compiler flag `-Wno-error=enum-constexpr-conversion` avoids the error, but then gives the wrong result: `1` rather than `0` is output.)

This contrasts with other undefined behavior: for example, changing the 5th line to `template <int V, int X = V * V>` shows that signed integer overflow is not a hard error but is SFINAE-able.

**Comment by shafik** [comment]
I am looking into this. We previously did not catch this undefined behavior in constant expression but https://reviews.llvm.org/D130058 changed that. I modified this error to be a warning due to the fallout this diagnostic had in a many code bases. So we now have a transition period before it turns into a hard error.

In order to do that we are not using the `CCEDiag` mechanism which feeds Notes into various parts of the code base. In this case it would be `Sema::CheckTemplateArgument(...)`. 

There may not be an easy way to make this SFINAE friendly until we turn it back into a hard error but I will see what I can do.

**Comment by dwblaikie** [comment]
are you sure this is a thing that should be SFINAE? Looks like it, but would be good to have some words from the spec that confirm it.

**Comment by AaronBallman** [comment]
> are you sure this is a thing that should be SFINAE? Looks like it, but would be good to have some words from the spec that confirm it.

http://eel.is/c++draft/temp.deduct#general-8 the "or expression" part is what makes this a SFINAE situation.

**Comment by dwblaikie** [comment]
> > are you sure this is a thing that should be SFINAE? Looks like it, but would be good to have some words from the spec that confirm it.
> 
> http://eel.is/c++draft/temp.deduct#general-8 the "or expression" part is what makes this a SFINAE situation.

ah, thanks!

**Comment by shafik** [comment]
PR: https://reviews.llvm.org/D150226
